packages/backend/src/open_payments/auth/middleware.ts

```diff
@@ -33,6 +34,13 @@ export function createAuthMiddleware({
       ) {
         ctx.throw(403, 'Insufficient Grant')
       }
+      await Grant.query()
+        .insert({
+          id: grant.grant,
+          clientId: grant.clientId
+        })
+        .onConflict('id')
+        .ignore()
```

Comment:
Very good questions.

> Should it only happen for the create incoming and outgoing payment routes? 
I think that is worth exploring.

>Should it happen inside a knex transaction that's passed to the route/service via ctx so that we don't actually insert if the route operation fails?

I don't see the risk of a grant being stored but the payment not being created (only tiny storage).